common/achievements.cpp

```diff
@@ -0,0 +1,178 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+
+#include "common/achievements.h"
+#include "common/config-manager.h"
+#include "common/debug.h"
+
+namespace Common {
+
+DECLARE_SINGLETON(AchievementsManager);
+
+
+AchievementsManager::AchievementsManager() {
+	unsetActiveDomain();
+}
+
+
+AchievementsManager::~AchievementsManager() {
+}
+
+bool AchievementsManager::setActiveDomain(AchievementsPlatform platform, const String &appId) {
+	String suffix = platform == STEAM_ACHIEVEMENTS ? "steam-" + appId : appId;
+	String achDomainId = "achievements-" + suffix;
+	String statDomainId = "statistics-" + suffix;
+
+	if (_achDomainId == achDomainId && _statDomainId == statDomainId) {
+		return true;
+	}
+
+	if (isReady()) {
+		unsetActiveDomain();
+	}
+
+	ConfMan.addMiscDomain(achDomainId);
+	ConfMan.addMiscDomain(statDomainId);
+
+	_achDomainId = achDomainId;
+	_statDomainId = statDomainId;
+	return true;
+}
+
+
+bool AchievementsManager::unsetActiveDomain() {
+	if (!isReady()) {
+		return true;
+	}
+
+	ConfMan.flushToDisk();
+
+	_achDomainId = "";
+	_statDomainId = "";
+	return true;
+}
+
+
+bool AchievementsManager::setAchievement(const String &id) {
+	if (!isReady()) {
+		return false;
+	}
+	if (isAchieved(id)) {
+		return true;
+	}
+
+	//TODO: insert a GUI notification here
+	debug("Achievement unlocked: %s", id.c_str());
+
+	ConfMan.setBool(id, true, _achDomainId);
+	ConfMan.flushToDisk();
```

Comment:
Saving the configuration file takes noticeable time on platforms with slow file IO, like the 3DS. However it's probably fine as it's not like achievements are unlocked every few seconds.

Reply:
I'm afraid there is no good solution for this.
Loosing displayed achievements is a very unpleasant scenario, so we need to save them ASAP.

There may exist some games that award player with several achievements simultaniously at certain moments, but it tend to happen on chapter/game end, when a pause occurs anyway (e.g. Completed the game + Completed the game in 1 hour + Completed the game without killing anyone). So I guess it's not a performance-critical scenario.

Reply:
There's the possibility of saving the achievements as per target save files instead of in the configuration file. The files would be smaller and faster to write. This would also have the benefit of synchronizing the achievements between devices over the cloud sync feature. The pull request is fine as it is for me, I'm just sharing the idea.

Reply:
That's a very good idea, actually. I have never used cloud sync yet, so I totally forgot to consider cloud sync scenarios.

There are a couple of things to thinks about though:
1. There is a pair of games that share a single achievements list ("JULIA: Among the Stars" and "JULIA: Untold"). You collect some achievements in main Wintermute game, and some other achievements in DLC, which is basically an additional Wintermute game. I wanted to keep original user experience of never getting 100% without playing both. I see no solutions other than identify data with platform id (e.g. "steam-257690"), not with target id (e.g. "juliauntold-win-1").
2. Merging mechanism would be nice for situations when achievement A was unlocked on device 1 and achievement B was unlocked on device 2. Simpliest way is to create a separate file for every unlocked achievement, but that means trashing Saves folder with hundreds of empty files with absurdly long filenames (e.g. game "steam-270570" has achievement "ACHIEVEMENT_HABLAR_CON_CHICA_ANTES_DE_DARLE_LA_FOTO"), which is not neat. And IIRC some backends are unable to create folders, which sadly means that "folder per platform id, file per achievement" is not an option.

Reply:
The idea with storing achievements into a separate file in the save directory is nice and relatively easily achievable.

How I would implement #1: Add another key to the game entry in the `scummvm.ini`, `achievements_file=<filename>`, and then for the games which share achievements, have it generated as the same filename, perhaps based on a target. 

#2: could work automagically with cloud support for saves. We already have mechanism for tracking save dates. So, you play on device A, achievements are stored and automatically synced to the cloud. You pick device B and it syncs achievements file from the cloud. PROFIT.

Reply:
Of course, for #1 you will have to read that file on options dialog opening.

Reply:
Wow, this was really easy. Switched AchMan from ConfMan to INIFile at 593f814c118c9cf34a449d0e1acfd8863c3e226d.

As an unexpected side effect of using `INIFile::saveToSaveFile()`, achievements data is now binary encoded, so now it's much more difficult to cheat.

Reply:
> Wow, this was really easy. Switched AchMan from ConfMan to INIFile at 593f814.

Awesome.

A small precision, the save sync does not handle conflicts as it is. In the situation where the local file has changed since the last sync and there is a new version available on the remote storage, whichever file has the newest timestamp will silently overwrite the other. Ideally the user would be asked which file they want to keep.

Reply:
Yes, this is how we implemented it, we had a discussion back then. E.g. if you play on device A in offline mode and get achievements X+Y, then play on device B in online (obviously it will not see these achievements), and get only achievement X, then if launched on device A in online mode, the file will be overwritten because save from device B will have a newer timestamp.

We considered it an acceptable compromise.

Though, displaying a prompt, or even make an option: "use server/use local/use newer timestamp(default)/ask" would close this once for all.